packages/blade/src/components/Button/BaseButton/BaseButton.tsx

```diff
@@ -0,0 +1,97 @@
+import type { ReactElement } from 'react';
+import getIn from '../../../utils/getIn';
+import BaseText from '../../Typography/BaseText';
+import { useTheme } from '../../BladeProvider';
+import type { IconComponentType } from '../../Icons';
+import StyledBaseButton from './StyledBaseButton';
+
+export type BaseButtonIcon = undefined | null | IconComponentType;
+
+type BaseButtonCommonProps = {
+  variant?: 'primary' | 'secondary' | 'tertiary' | 'link';
+  intent?: 'positive' | 'negative' | 'notice' | 'info' | 'neutral';
+  contrast?: 'low' | 'high';
+  size?: 'large' | 'medium' | 'small' | 'xsmall';
+  iconPosition?: 'left' | 'right';
+  isDisabled?: boolean;
+  isFullWidth?: boolean;
+  onClick?: () => void;
+  type?: 'button' | 'reset' | 'submit';
+};
+
+type BaseButtonWithoutIconProps = BaseButtonCommonProps & {
+  icon?: undefined;
+  children: string;
+};
+
+type BaseButtonWithIconProps = BaseButtonCommonProps & {
+  icon: IconComponentType;
+  children?: string;
+};
+
+export type BaseButtonProps<T> = T extends {
+  icon?: infer Icon;
+}
+  ? Icon extends Extract<BaseButtonIcon, IconComponentType>
```

Comment:
why extract?

Reply:
Wanted it to represent that `Icon` needs to be one of the types from `BaseButtonIcon`. I could keep it `Icon extends IconComponentType` but it seems very detached from the `BaseButtonIcon`. Imagine `IconComponentType` evolves over time but we fail to change it here